[FEATURE] Drive Identification and Refactoring

ukos-git opened the pull request:
# Description
- USB devices frequently change their mount points on system reboot, causing faulty job associations to drives. To prevent this, drive recognition based on pyudev serial information (maker, model, serial number) of optical drives is introduced.
- MakeMKV info calls create zombie processes and syslog errors at MakeMKV level. These changes help to reduce the amount of MakeMKV info calls getting in each other's way.

## Type of change

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

# How Has This Been Tested?

- [X ] Docker

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have tested that my fix is effective or that my feature works

# Changelog:

- Handle race condition of concurrent MakeMKV info calls creating zombie processes and silently failing.
- Enhance MakeMKV Calls with Parser and Error Handling.
- Reduce MakeMKV info calls by caching the MakeMKV drive id on calls to disc:9999
- Handle alternating mount points for USB devices.
- Add support for displaying more device info
- On system startup, scan drives to recognize hardware changes. Mark unfound devices as stale and delete their last mount point.
- The layout of the drive tab was changed to include more information on the drives.
- DataClasses and Enums are used for better usability of the MakeMKV and pyudev output.

# Logs

[MONSTERS_INC_DEU.log.gz](https://github.com/user-attachments/files/18086423/MONSTERS_INC_DEU.log.gz)
[FINDING_NEMO.log.gz](https://github.com/user-attachments/files/18086424/FINDING_NEMO.log.gz)

microtechno9000 commented:
Thanks for submitting a PR. I havent been able to delve through and review the changes yet. Will have a look over the next week.

Can you look at the Sonarcube results and resolve any issues it highlights.

ukos-git commented:
I'd need some help with the SonarCloud issues. From my perspective, they are of "wont-fix" type. I guess, the code was not on A rating before I touched it. I'd consider "B" an improvement. Less refactoring means less SonarCloud Issues.

I recommend to keep the PR open for a few more days since I'd like to check the parser with a few more discs (Currently have around 20 rips completed but all in mkv mode).

microtechno9000 commented:
Agree, the ripper code is quite bad and it is on the rework path.

If there are issues that Sonarcloud raises that are existing just mark them as that. Unless there is an easy fix

ukos-git commented:
I worked around most of the issues but the [last one is tricky](https://sonarcloud.io/project/issues?pullRequest=1259&open=AZOymfXOeyeAJeRfgb0P&id=automatic-ripping-machine_automatic-ripping-machine) and wants me to further refactor `get_track_info`. With the class based approach I could merge a lot of functions into one seemingly simple construct. I understand that nesting is bad but in this case I think it looks ok. I initially had switch statements here which makes pylint happy but is not python 3.6 compatible. I am thinking of a coroutine but maybe not today.

```python
def get_track_info(index, job):                                                 
    """                                                                         
    Use MakeMKV to get track info and update Track class                        
                                                                                
    Parameters:                                                                 
        index: Makemkv disc index                                               
        job: arm.models.job.Job                                                              
    Returns:                                                                    
        None                                                                    
                                                                                
    .. note:: For help with MakeMKV codes:                                      
    https://github.com/automatic-ripping-machine/automatic-ripping-machine/wiki/MakeMKV-Codes
    """                                                                         
                                                                                
    logging.info("Using MakeMKV to get information on all the tracks on the disc. This will t
    options = [                                                                 
        f"--progress={progress_log(job)}",                                      
        f"--minlength={job.config.MINLENGTH}",                                  
    ]                                                                           
    track_id = 0                                                                
    fps = float(0)                                                              
    aspect = ""                                                                 
    seconds = 0                                                                 
    filename = ""                                                               
    stream_type = MAKEMKV_STREAM_CODE_TYPE_VIDEO  # save to assume for track_id 0.
    output_types = (                                                            
        OutputType.CINFO |                                                      
        OutputType.SINFO |                                                      
        OutputType.TCOUNT |                                                     
        OutputType.TINFO                                                        
    )                                                                           
    for message in makemkv_info(job, select=output_types, index=index, options=options):
        if isinstance(message, (TInfo, SInfo)) and track_id != message.tid:     
            # Next track was detected. Add current.                             
            utils.put_track(job, track_id, seconds, aspect, str(fps), False, SOURCE, filename
            track_id = message.tid                                              
        if isinstance(message, SInfo):                                          
            assert message.tid == track_id, message                             
            if message.id == StreamID.TYPE:                                     
                stream_type = message.code                                      
            elif stream_type == MAKEMKV_STREAM_CODE_TYPE_VIDEO:                 
                if message.id == StreamID.ASPECT:                               
                    aspect = message.value.strip()                              
                elif message.id == StreamID.FPS:                                
                    fps = float(message.value.split()[0])                       
        elif isinstance(message, TInfo):                                        
            assert message.tid == track_id, message                             
            if message.id == TrackID.FILENAME:                                  
                filename = message.value.strip()
            elif message.id == TrackID.DURATION:                                
                seconds = convert_to_seconds(message.value.strip())             
        elif isinstance(message, Titles):                                       
            logging.info(f"Found {message.count:d} titles")                     
            utils.database_updater({"no_of_titles": message.count}, job)        
    # If we haven't already added any tracks add one with what we have                       
    utils.put_track(job, track_id, seconds, aspect, str(fps), False, SOURCE, filename)
```

Would be good to have the new parsers for pyudev and makemkv also in the development branches for the next release. Would be eager to help.

ukos-git commented:
this is what chatGPT came up with after some questioning. Haven't tested it but looks good at first glance:

```python
import asyncio
from dataclasses import dataclass
from typing import AsyncIterator

@dataclass
class TrackInfo:
    track_id: int
    aspect: str = ""
    fps: float = 0.0
    filename: str = ""
    seconds: int = 0

async def process_message_field(messages: AsyncIterator, expected_class, field_id, parse=lambda x: x.strip()):
    """Process specific fields from messages."""
    async for message in messages:
        if isinstance(message, expected_class) and message.id == field_id:
            yield parse(message.value)

async def dispatcher(messages: AsyncIterator, track_id: int, coroutines):
    """Send messages with the same track_id to the relevant coroutines."""
    async for message in messages:
        if hasattr(message, 'tid') and message.tid == track_id:  # Check for track ID match
            for coro in coroutines:
                await coro.asend(message)

async def get_track_info(index, job):
    """Use MakeMKV to get track info and update Track class"""
    logging.info("Using MakeMKV to get information on all the tracks on the disc.")

    options = [
        f"--progress={progress_log(job)}",
        f"--minlength={job.config.MINLENGTH}",
    ]

    output_types = (
        OutputType.CINFO |
        OutputType.SINFO |
        OutputType.TCOUNT |
        OutputType.TINFO
    )

    messages = makemkv_info(job, select=output_types, index=index, options=options)

    async for message in messages:
        if isinstance(message, Titles):
            logging.info(f"Found {message.count:d} titles")
            utils.database_updater({"no_of_titles": message.count}, job)

        if isinstance(message, TInfo) and message.id == TrackID.TRACK_ID:
            track_id = message.tid
            track_info = TrackInfo(track_id=track_id)

            aspect = process_message_field(messages, SInfo, StreamID.ASPECT)
            fps = process_message_field(messages, SInfo, StreamID.FPS, parse=float)
            filename = process_message_field(messages, TInfo, TrackID.FILENAME)
            seconds = process_message_field(messages, TInfo, TrackID.DURATION, parse=convert_to_seconds)

            await aspect.asend(None)
            await fps.asend(None)
            await filename.asend(None)
            await seconds.asend(None)

            await dispatcher(messages, track_id, [aspect, fps, filename, seconds])

            track_info.aspect = await anext(aspect, "")
            track_info.fps = await anext(fps, 0.0)
            track_info.filename = await anext(filename, "")
            track_info.seconds = await anext(seconds, 0)

            utils.put_track(job, track_info.track_id, track_info.seconds, track_info.aspect, str(track_info.fps), False, SOURCE, track_info.filename)

    # Ensure the last track is added if any
    if 'track_info' in locals():
        utils.put_track(job, track_info.track_id, track_info.seconds, track_info.aspect, str(track_info.fps), False, SOURCE, track_info.filename)
```

microtechno9000 commented on arm/models/__init__.py:
Why add this, it means all the models are imported when importing the models. Currently we only import the modules we need, when required.

ukos-git replied:
From my point, submodules should expose their "Interfaces" directly. This ensures that there are no further dependencies throughout the code that use deeper nested functionality. It also avoids calls like "models.job.Job". It is for better readability and maintainability. I agree that this suffers from loading performance and RAM usage, but the model classes should not add too much overhead. I am good without it, though,  if you think this is not what we want here.

microtechno9000 replied:
Fine with the models being autoloaded

microtechno9000 commented:
Thanks for looking at the Sonarcube issues, they look good.
I should have mentioned we are working on a ripper logic re-write with v3.x because it is too hard to add new features and capabilities. If you have spare time you are welcome to pull the [3.0_devel branch](https://github.com/automatic-ripping-machine/automatic-ripping-machine/tree/3.0_devel)

Can you please look at the below bugs

1. Loading index post the database migrate throws an error, unfortunately not very descriptive.
```bash
[2024-12-26 10:53:22,792] INFO ARM: utils.arm_db_migrate ARM head: 5ceacc391408 database: 5ceacc391408
[2024-12-26 10:53:22,792] INFO ARM: utils.arm_db_migrate Database is now up to date
[2024-12-26 10:53:22,810] DEBUG ARM: routes.handle_exception Error: encoding without a string argument
[2024-12-26 10:53:22,814] DEBUG ARM: notifications.arm_nav_notify 0
```
2. Loading the settings page also throws an error
```bash
[2024-12-26 10:57:30,578] DEBUG ARM: DriveUtils.update_job_status Drive Mode: auto
[2024-12-26 10:57:30,578] DEBUG ARM: DriveUtils.update_job_status Drive Mode: auto
[2024-12-26 10:57:30,587] DEBUG ARM: notifications.arm_nav_notify 0
[2024-12-26 10:57:30,604] DEBUG ARM: routes.handle_exception Error: 'None' has no attribute 'upper'
[2024-12-26 10:57:30,605] DEBUG ARM: notifications.arm_nav_notify 0

```

ukos-git commented:
I admit that I got carried away with refactoring. I will rebase to v3 branch.

I have added the exception trace info to the routes.handle_exception logger for better debugging.

Looking at the errors, I would assume that you are missing some fields in your pyudev.Context() output. That is probably why the pyudev info is currently spammed into the log files for every rip. Is there by chance a documentation on the fields that are available for *every* optical device? - Otherwise we would need to live with potential None values here. There should be at lease be "ID_SERIAL" for every device for my idea of not taking the mount point to work properly.

microtechno9000 commented:
> I admit that I got carried away with refactoring. I will rebase to v3 branch.
> 
> I have added the exception trace info to the routes.handle_exception logger for better debugging.
> 
> Looking at the errors, I would assume that you are missing some fields in your pyudev.Context() output. That is probably why the pyudev info is currently spammed into the log files for every rip. Is there by chance a documentation on the fields that are available for *every* optical device? - Otherwise we would need to live with potential None values here. There should be at lease be "ID_SERIAL" for every device for my idea of not taking the mount point to work properly.

Wow, quick fix

Don't need to rebase all your changes. Just  if you are spending too much time on the reactor it might be lost with the update to v3.0

I can have a look at what pyudev outputs, but from experience have had issues with some of these sub modules not generating  standards output. lmsensors is great for this between CPU types.

microtechno9000 commented:
Checked out the changes

1. Crash/error on loading has gone
2. Crash/error on loading settings is gone
3. Love that the settings page has been fixed with the current drive open/closed state, that has been on my list to fix for a while, since #695 
4. Please put the name of the drive back, the name and description are useful for users (selfishly I like the name)
5. When scanning for new drives and using DEBUG, the logs have been removed. They are there for developers when debugging, please put them back. Also useful when users have an issue, can provide support via Github or discord with those logs
6. From the settings page, using the open/close drive icon only opens the drive, I was unable to close the drive

both drives open
```bash
[2024-12-28 10:46:10,793] DEBUG ARM: system_drives.tray_status Drive '/dev/sr1': tray status 'CDS.TRAY_OPEN'
[2024-12-28 10:46:10,853] DEBUG ARM: system_drives.tray_status Drive '/dev/sr0': tray status 'CDS.TRAY_OPEN'
```

Manually closing the status updates
```bash
[2024-12-28 10:47:58,646] DEBUG ARM: system_drives.tray_status Drive '/dev/sr1': tray status 'CDS.NO_DISC'
[2024-12-28 10:47:58,710] DEBUG ARM: system_drives.tray_status Drive '/dev/sr0': tray status 'CDS.NO_DISC'
```

ukos-git commented:
Thank you for the review. I moved the drive Information to the db class and will put that back to log to stdout. Naming comes also back. It was just an easy way out.

Regarding open/close i did consider that this is nothing to be stored in db. Most drives have an eject button. In fact, i could not test that brach since my drives only eject and dont close at all. 

Give me some time to rebase and I will be happy to tackle these issues.

microtechno9000 commented:
Thanks, that would be good

The commands that have been working are 

Eject drive
`eject sr0` 

Close drive
`eject -t sr0`

simonmcnair commented:
did you try eject -T sr0 ?  I've been using the toggle and it works quite well

ukos-git commented:
I fixed a few issues with my code here and there, re-added the name, used the toggle command, added output of eject to the logs etc. On my way, I also came up with a simple solution for [Add a "Ripping Complete, Safe to Eject" So you can start ripping a second disk while the first is still transcoding #1236](https://github.com/automatic-ripping-machine/automatic-ripping-machine/issues/1236#issue-2642938308).

The extended drive information is working flawlessly on my end. Would be interesting to know for example, if the `ID_VENDOR_ENC` and `ID_MODEL_ENC` is available on all systems. A fallback could use `ID_MODEL` but overall this information is purely cosmetic.

The changes to the blocking calls of makemkvcon info work. I don't get hanging makemkv info processes on my machine anymore. Unfortunately, I cannot control priority of info over backup/mkv, so sometimes the long lasting makemkvcon rip takes over and blocks the slot for further info processes. For this version it is fine but I consider reworking this once more for v3 and add a queueing system which would solve 3 currently open issues as mentioned in #1263 

I'd like to close this PR here for v2 and start working on v3 with a new PR. I have certainly overdone my refactoring here. - 

On a short note: A testing system with mock output of different setups with pyudev and makemkv output would be nice to have.

microtechno9000 commented:
Thanks, I will have a look next week. I am on holidays at the moment

microtechno9000 commented:
@ukos-git Please resolve the merge conflicts and per my previous comments, please add logs showing the changes work through ARM

Unfortunately we don't have unit-testing, or end-to-end testing that we can run yet to confirm changes work.
The only way to confirm is to run a movie through ARM and check it works

ukos-git commented:
I added a mask to the serial number output of the drives. Serial numbers are, however, still printed by Udev output. But the following log was done with the recent changes (had to build arm docker myself to include the most recent [MakeMKV 1.17.9](https://forum.makemkv.com/forum/viewtopic.php?f=3&t=224)

[CRANK_BD_173906217445.log](https://github.com/user-attachments/files/18724577/CRANK_BD_173906217445.log)

I have found an issue that probably has to do with the way, backpopulate works between the tables Job and Drive but I can only reproduce it on failed "makemkv info" processes but because I touched the backpopulate in my changes, I will look into it.

[THE_LEGEND_OF_HERCULES_173906160365.log](https://github.com/user-attachments/files/18724590/THE_LEGEND_OF_HERCULES_173906160365.log)

ukos-git commented:
@microtechno9000 did I miss anything?

microtechno9000 commented:
looks like its all there

will have a look through the code and logs in the next few days

microtechno9000 commented:
@ukos-git Sorry it has been a while, I thought I was still waiting info for this PR. I see you provided it all back in Feb and I didn't remove the label for my checks.

Can you resolve the conflicts against this PR, once they are in I can merge. I don't see any issues with the PR.

ukos-git commented:
@microtechno9000 This is ready for review. I added a few more commits starting with [286de8b](https://github.com/automatic-ripping-machine/automatic-ripping-machine/pull/1259/commits/286de8b13c89151bf69636a1493197c6be2f4707). Let me know if this is ok to add. Currently I am getting MakeMKV shareware expired errors. These Errors and some (hard to read) discs that I used for testing the PR were a good test for the new MakeMKV parser since now we can raise these errors instead of proceeding to HandBrake (unaware that there were no mkv files produced by makemkv). Raising is needed since the exit code of MakeMKV is unreliable (mostly 0). I also tackled some Sonar Cube issues on the way. 

Please give me some time for further functional testing. I need to check that the new changes to JobState work in any case and that rebasing did not destroy the changes.
I will check this PR in a few weeks again.

microtechno9000 commented:
@ukos-git I have reviewed the code and the changes are great, a lot more error checking, catches for odd MakeMKV states and better handling of the drives
If you are ready, I can approve and merge the changes in

microtechno9000 commented:
Tested locally with a Bluray and can confirm changes are good
merging into main. Version increment and ARM Wiki failed jobs will be resolved separately
